debian/rtl_eeprom.1
.TH "rtl_eeprom" 1 "0.5.0" RTL-SDR "User Commands"
.SH NAME
rtl-eeprom \- EEPROM programming tool for RTL2832 based DVB-T receivers
.SH DESCRIPTION
Dumps configuration and also writes EEPROM configuration.
Written by Steve Markgraf and incorporated in the osmocom rtl-sdr project.
.LP
Use at your own risk, especially -w!
.LP
Much software is available for the RTL2832. Most of the user-level
packages rely on the librtlsdr library which comes as part of the
rtl-sdr codebase. This codebase contains both the library itself and
also a number of command line tools such as rtl_test, rtl_sdr,
rtl_tcp, and rtl_fm. These command line tools use the library to test
for the existence of RTL2832 devices and to perform basic data
transfer functions to and from the device.
.LP
Because most of the RTL2832 devices are connected using USB, the
librtlsdr library depends on the libusb library to communicate with
the device.
.SH USAGE
Writing bad information to the EEPROM will make your
device useless. 
.SH SYNOPSIS
.B  rtl_eeprom [OPTIONS]
.SH OPTIONS
.IP "-d device_index (default: 0)"
.IP "-m <str> set manufacturer string"
.IP "-p <str> set product string"
.IP "-s <str> set serial number string"
.IP "-i <0,1> disable/enable IR-endpoint"
.IP "-g <conf> generate default config and write to device"
   <conf> can be one of:
   realtek  Realtek default (as without EEPROM)
   realtek_oem  Realtek default OEM with EEPROM
   noxon  Terratec NOXON DAB Stick
   terratec_black  Terratec T Stick Black
   terratec_plus  Terratec T Stick+ (DVB-T/DAB)
.IP "-w <filename> write dumped file to device"
.IP "-r <filename> dump EEPROM to file"
.IP "-h display this help text"
.LP
Use on your own risk, especially -w!
.SH SEE ALSO
RTL-SDR wiki documentation:
.B http://sdr.osmocom.org/trac/wiki/rtl-sdr
.LP
Other rtl-sdr programs:
.sp
rtl_adsb(1), rtl_fm(1), rtl_sdr(1), rtl_tcp(1), rtl_test(1)
.SH AUTHOR
This manual page was written by Maitland Bottoms
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2013 A. Maitland Bottoms <bottoms@debian.org>
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


debian/changelog
rtl-sdr (0.6git) unstable; urgency=medium

  * New upstream release

 -- Harald Welte <laforge@gnumonks.org>  Sun, 06 Jun 2018 15:09:42 +0200

rtl-sdr (0.5.4-1) unstable; urgency=medium

  * New upstream release
  * update to v0.5.4-3-ga854ae8
    use USB zero-copy transfers if possible

 -- A. Maitland Bottoms <bottoms@debian.org>  Sat, 12 May 2018 16:49:43 -0400

rtl-sdr (0.5.3-14) unstable; urgency=medium

  * update to v0.5.3-20-g4520f00 (Closes: #892974)
  * minimal ipv6 support (Closes: #870804)
  * VCS to salsa
  * AppStream metadata.xml

 -- A. Maitland Bottoms <bottoms@debian.org>  Mon, 16 Apr 2018 20:45:53 -0400

rtl-sdr (0.5.3-13) unstable; urgency=medium

  * build with libusb-1.0-0-dev stub on hurd-i386
  * initial ipv6 support for rtl_tcp...

 -- A. Maitland Bottoms <bottoms@debian.org>  Thu, 23 Nov 2017 15:59:40 -0500

rtl-sdr (0.5.3-12) unstable; urgency=medium

  * add new HanfTek dongle
  * Bias T support (Closes: #854378, #842249)

 -- A. Maitland Bottoms <bottoms@debian.org>  Wed, 23 Aug 2017 23:31:27 -0400

rtl-sdr (0.5.3-11) unstable; urgency=medium

  * correct invocation of rm_conffile (Thanks Chris!) (Closes: #838161)
  * drop uploaders line on advice of MIA team. (Closes: #836590)

 -- A. Maitland Bottoms <bottoms@debian.org>  Sat, 08 Oct 2016 11:17:47 -0400

rtl-sdr (0.5.3-10) unstable; urgency=medium

  * remove rtl-sdr-blacklist.conf on upgrade. Thanks Bob! (Closes: #829517)

 -- A. Maitland Bottoms <bottoms@debian.org>  Sat, 09 Jul 2016 23:38:24 -0400

rtl-sdr (0.5.3-9) unstable; urgency=medium

  * Edit of debian/librtlsdr0.udev in 0.5.3-8 was a no-op. Real fix
    done to debian/patches/use-udev-uaccess-rules. (Closes: #825073)

 -- A. Maitland Bottoms <bottoms@debian.org>  Wed, 25 May 2016 17:19:57 -0400

rtl-sdr (0.5.3-8) unstable; urgency=high

  * Fix syntax errors for systemd-udevd in udev rules (Closes: #825073)

 -- A. Maitland Bottoms <bottoms@debian.org>  Tue, 24 May 2016 21:08:50 -0400

rtl-sdr (0.5.3-7) unstable; urgency=medium

  * better udev rules (more like camera devices in libgphoto2-6)

 -- A. Maitland Bottoms <bottoms@debian.org>  Tue, 10 May 2016 19:20:27 -0400

rtl-sdr (0.5.3-6) unstable; urgency=medium

  * Use ENV{ID_SOFTWARE_RADIO}=1 in udev rules (Closes: #823089)
  * Enable DETACH_KERNEL_DRIVER (Closes: 823022)
  * Make myself maintainer so I get the bug reports

 -- A. Maitland Bottoms <bottoms@debian.org>  Sun, 08 May 2016 12:12:13 -0400

rtl-sdr (0.5.3-5) unstable; urgency=medium

  * Add watch fiule
  * place rtl-sdr -n comm section (Closes: #758077)
  * improve-librtlsdr-pc-file (Closes: #784912)
  * improve-scanning-range-parsing (LP: #1469478)

 -- A. Maitland Bottoms <bottoms@debian.org>  Sun, 23 Aug 2015 10:35:42 -0400

rtl-sdr (0.5.3-4) unstable; urgency=low

  * Update to v0.5.3-12-ge3c03f7

 -- A. Maitland Bottoms <bottoms@debian.org>  Sat, 08 Aug 2015 23:43:54 -0400

rtl-sdr (0.5.3-3) unstable; urgency=low

  * Update to v0.5.3-5-g6ee5573
    lib: handle events after canceling transfers
    lib: change default number of transfers to 15
    rtl_tcp: make all global variables static

 -- A. Maitland Bottoms <bottoms@debian.org>  Sun, 13 Apr 2014 10:48:49 -0400

rtl-sdr (0.5.3-2) unstable; urgency=low

  * Upstream: lib: only print to stderr in tuner_r82xx_set_pll()
  * Update man pages (New -M (modulation) and -E (option) and ppm setting)
  * Have librtlsdr0 also install a blacklist for linux video drivers

 -- A. Maitland Bottoms <bottoms@debian.org>  Sat, 08 Feb 2014 22:40:06 -0500

rtl-sdr (0.5.3-1) unstable; urgency=low

  * New upstream git tag release

 -- A. Maitland Bottoms <bottoms@debian.org>  Thu, 06 Feb 2014 20:45:38 -0500

rtl-sdr (0.5.2.7.3ab6-1~bpo70+1) wheezy-backports; urgency=low

  * Rebuild for wheezy-backports.

 -- A. Maitland Bottoms <bottoms@debian.org>  Tue, 21 Jan 2014 19:34:16 -0500

rtl-sdr (0.5.2.7.3ab6-1) unstable; urgency=low

  * New upstream snapshot

 -- A. Maitland Bottoms <bottoms@debian.org>  Sun, 29 Dec 2013 21:37:19 -0500

rtl-sdr (0.5.1.14.360d-1~wheezy) stable; urgency=low

  * New upstream snapshot
  * GNU Radio LiveDVD 2013-1110

 -- A. Maitland Bottoms <bottoms@debian.org>  Mon, 11 Nov 2013 12:46:00 -0500

rtl-sdr (0.5.0.4.4914-2) unstable; urgency=low

  * Use kfreebsd libusb

 -- A. Maitland Bottoms <bottoms@debian.org>  Fri, 01 Nov 2013 17:16:42 -0400

rtl-sdr (0.5.0.4.4914-1) stable; urgency=low

  * New upstream snapshot (Closes: #701018).
  * Match GNU Radio live distribution version
  * Sponsored upload

 -- A. Maitland Bottoms <bottoms@debian.org>  Sat, 28 Sep 2013 16:55:08 -0400

rtl-sdr (0.5.0+git20130715-1) unstable; urgency=low

  * Initial release (Closes: #701018).

 -- Adam Cécile (Le_Vert) <gandalf@le-vert.net>  Mon, 15 Jul 2013 15:51:05 +0200

librtlsdr (0.0git3198f14-1) unstable; urgency=low

  * New upstream git

 -- A. Maitland Bottoms <bottoms@debian.org>  Mon, 14 May 2012 20:28:18 -0400


src/rtl_eeprom.c
/*
 * rtl-sdr, turns your Realtek RTL2832 based DVB dongle into a SDR receiver
 * rtl_eeprom, EEPROM modification tool
 * Copyright (C) 2012 by Steve Markgraf <steve@steve-m.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#ifndef _WIN32
#include <unistd.h>
#else
#include <windows.h>
#include "getopt/getopt.h"
#endif

#include "rtl-sdr.h"

#define EEPROM_SIZE	256
#define MAX_STR_SIZE	256
#define STR_OFFSET	0x09

static rtlsdr_dev_t *dev = NULL;

typedef struct rtlsdr_config {
	uint16_t vendor_id;
	uint16_t product_id;
	char manufacturer[MAX_STR_SIZE];
	char product[MAX_STR_SIZE];
	char serial[MAX_STR_SIZE];
	int have_serial;
	int enable_ir;
	int remote_wakeup;
} rtlsdr_config_t;

void dump_config(rtlsdr_config_t *conf)
{
	fprintf(stderr, "__________________________________________\n");
	fprintf(stderr, "Vendor ID:\t\t0x%04x\n", conf->vendor_id);
	fprintf(stderr, "Product ID:\t\t0x%04x\n", conf->product_id);
	fprintf(stderr, "Manufacturer:\t\t%s\n", conf->manufacturer);
	fprintf(stderr, "Product:\t\t%s\n", conf->product);
	fprintf(stderr, "Serial number:\t\t%s\n", conf->serial);
	fprintf(stderr, "Serial number enabled:\t");
	fprintf(stderr, conf->have_serial ? "yes\n": "no\n");
	fprintf(stderr, "IR endpoint enabled:\t");
	fprintf(stderr, conf->enable_ir ? "yes\n": "no\n");
	fprintf(stderr, "Remote wakeup enabled:\t");
	fprintf(stderr, conf->remote_wakeup ? "yes\n": "no\n");
	fprintf(stderr, "__________________________________________\n");
}

void usage(void)
{
	fprintf(stderr,
		"rtl_eeprom, an EEPROM programming tool for "
		"RTL2832 based DVB-T receivers\n\n"
		"Usage:\n"
		"\t[-d device_index (default: 0)]\n"
		"\t[-m <str> set manufacturer string]\n"
		"\t[-p <str> set product string]\n"
		"\t[-s <str> set serial number string]\n"
		"\t[-i <0,1> disable/enable IR-endpoint]\n"
		"\t[-g <conf> generate default config and write to device]\n"
		"\t[   <conf> can be one of:]\n"
		"\t[   realtek\t\tRealtek default (as without EEPROM)]\n"
		"\t[   realtek_oem\t\tRealtek default OEM with EEPROM]\n"
		"\t[   noxon\t\tTerratec NOXON DAB Stick]\n"
		"\t[   terratec_black\tTerratec T Stick Black]\n"
		"\t[   terratec_plus\tTerratec T Stick+ (DVB-T/DAB)]\n"
		"\t[-w <filename> write dumped file to device]\n"
		"\t[-r <filename> dump EEPROM to file]\n"
		"\t[-h display this help text]\n"
		"\nUse on your own risk, especially -w!\n");
	exit(1);
}

int get_string_descriptor(int pos, uint8_t *data, char *str)
{
	int len, i, j = 0;

	len = data[pos];

	if (data[pos + 1] != 0x03)
		fprintf(stderr, "Error: invalid string descriptor!\n");

	for (i = 2; i < len; i += 2)
		str[j++] = data[pos + i];

	str[j] = 0x00;

	return pos + i;
}

int set_string_descriptor(int pos, uint8_t *data, char *str)
{
	int i = 0, j = 2;

	if (pos < 0)
		return -1;

	data[pos + 1] = 0x03;

	while (str[i] != 0x00) {
		if ((pos + j) >= 78) {
			fprintf(stderr, "Error: string too long, truncated!\n");
			return -1;
		}
		data[pos + j++] = str[i++];
		data[pos + j++] = 0x00;
	}

	data[pos] = j;

	return pos + j;
}

int parse_eeprom_to_conf(rtlsdr_config_t *conf, uint8_t *dat)
{
	int pos;

	if ((dat[0] != 0x28) || (dat[1] != 0x32))
		fprintf(stderr, "Error: invalid RTL2832 EEPROM header!\n");

	conf->vendor_id = dat[2] | (dat[3] << 8);
	conf->product_id = dat[4] | (dat[5] << 8);
	conf->have_serial = (dat[6] == 0xa5) ? 1 : 0;
	conf->remote_wakeup = (dat[7] & 0x01) ? 1 : 0;
	conf->enable_ir = (dat[7] & 0x02) ? 1 : 0;

	pos = get_string_descriptor(STR_OFFSET, dat, conf->manufacturer);
	pos = get_string_descriptor(pos, dat, conf->product);
	get_string_descriptor(pos, dat, conf->serial);

	return 0;
}

int gen_eeprom_from_conf(rtlsdr_config_t *conf, uint8_t *dat)
{
	int pos;

	dat[0] = 0x28;
	dat[1] = 0x32;
	dat[2] = conf->vendor_id & 0xff;
	dat[3] = (conf->vendor_id >> 8) & 0xff ;
	dat[4] = conf->product_id & 0xff;
	dat[5] = (conf->product_id >> 8) & 0xff;
	dat[6] = conf->have_serial ? 0xa5 : 0x00;
	dat[7] = 0x14;
	dat[7] |= conf->remote_wakeup ? 0x01 : 0x00;
	dat[7] |= conf->enable_ir ? 0x02 : 0x00;
	dat[8] = 0x02;

	pos = set_string_descriptor(STR_OFFSET, dat, conf->manufacturer);
	pos = set_string_descriptor(pos, dat, conf->product);
	pos = set_string_descriptor(pos, dat, conf->serial);

	dat[78] = 0x00;		/* length of IR config */

	return pos;
}

enum configs {
	CONF_NONE = 0,
	REALTEK,
	REALTEK_EEPROM,
	TERRATEC_NOXON,
	TERRATEC_T_BLACK,
	TERRATEC_T_PLUS,
};

void gen_default_conf(rtlsdr_config_t *conf, int config)
{
	switch (config) {
	case REALTEK:
		fprintf(stderr, "Realtek default (as without EEPROM)\n");
		conf->vendor_id = 0x0bda;
		conf->product_id = 0x2832;
		strcpy(conf->manufacturer, "Generic");
		strcpy(conf->product, "RTL2832U DVB-T");
		strcpy(conf->serial, "0");
		conf->have_serial = 1;
		conf->enable_ir = 0;
		conf->remote_wakeup = 1;
		break;
	case REALTEK_EEPROM:
		fprintf(stderr, "Realtek default OEM with EEPROM\n");
		conf->vendor_id = 0x0bda;
		conf->product_id = 0x2838;
		strcpy(conf->manufacturer, "Realtek");
		strcpy(conf->product, "RTL2838UHIDIR");
		strcpy(conf->serial, "00000001");
		conf->have_serial = 1;
		conf->enable_ir = 1;
		conf->remote_wakeup = 0;
		break;
	case TERRATEC_NOXON:
		fprintf(stderr, "Terratec NOXON DAB Stick\n");
		conf->vendor_id = 0x0ccd;
		conf->product_id = 0x00b3;
		strcpy(conf->manufacturer, "NOXON");
		strcpy(conf->product, "DAB Stick");
		strcpy(conf->serial, "0");
		conf->have_serial = 1;
		conf->enable_ir = 0;
		conf->remote_wakeup = 1;
		break;
	case TERRATEC_T_BLACK:
		fprintf(stderr, "Terratec T Stick Black\n");
		conf->vendor_id = 0x0ccd;
		conf->product_id = 0x00a9;
		strcpy(conf->manufacturer, "Realtek");
		strcpy(conf->product, "RTL2838UHIDIR");
		strcpy(conf->serial, "00000001");
		conf->have_serial = 1;
		conf->enable_ir = 1;
		conf->remote_wakeup = 0;
		break;
	case TERRATEC_T_PLUS:
		fprintf(stderr, "Terratec ran T Stick+\n");
		conf->vendor_id = 0x0ccd;
		conf->product_id = 0x00d7;
		strcpy(conf->manufacturer, "Realtek");
		strcpy(conf->product, "RTL2838UHIDIR");
		strcpy(conf->serial, "00000001");
		conf->have_serial = 1;
		conf->enable_ir = 1;
		conf->remote_wakeup = 0;
		break;
	default:
		break;
	};
}

int main(int argc, char **argv)
{
	int i, r, opt;
	uint32_t dev_index = 0;
	int device_count;
	char *filename = NULL;
	FILE *file = NULL;
	char *manuf_str = NULL;
	char *product_str = NULL;
	char *serial_str = NULL;
	uint8_t buf[EEPROM_SIZE];
	rtlsdr_config_t conf;
	int flash_file = 0;
	int default_config = 0;
	int change = 0;
	int ir_endpoint = 0;
	char ch;

	while ((opt = getopt(argc, argv, "d:m:p:s:i:g:w:r:h?")) != -1) {
		switch (opt) {
		case 'd':
			dev_index = atoi(optarg);
			break;
		case 'm':
			manuf_str = optarg;
			change = 1;
			break;
		case 'p':
			product_str = optarg;
			change = 1;
			break;
		case 's':
			serial_str = optarg;
			change = 1;
			break;
		case 'i':
			ir_endpoint = (atoi(optarg) > 0) ? 1 : -1;
			change = 1;
			break;
		case 'g':
			if (!strcmp(optarg, "realtek"))
				default_config = REALTEK;
			else if (!strcmp(optarg, "realtek_oem"))
				default_config = REALTEK_EEPROM;
			else if (!strcmp(optarg, "noxon"))
				default_config = TERRATEC_NOXON;
			else if (!strcmp(optarg, "terratec_black"))
				default_config = TERRATEC_T_BLACK;
			else if (!strcmp(optarg, "terratec_plus"))
				default_config = TERRATEC_T_PLUS;

			if (default_config != CONF_NONE)
				change = 1;
			break;
		case 'w':
			flash_file = 1;
			change = 1;
			/* fall-through */
		case 'r':
			filename = optarg;
			break;
		default:
			usage();
			break;
		}
	}

	device_count = rtlsdr_get_device_count();
	if (!device_count) {
		fprintf(stderr, "No supported devices found.\n");
		exit(1);
	}

	fprintf(stderr, "Found %d device(s):\n", device_count);
	for (i = 0; i < device_count; i++)
		fprintf(stderr, "  %d:  %s\n", i, rtlsdr_get_device_name(i));
	fprintf(stderr, "\n");

	fprintf(stderr, "Using device %d: %s\n",
		dev_index,
		rtlsdr_get_device_name(dev_index));

	r = rtlsdr_open(&dev, dev_index);
	if (r < 0) {
		fprintf(stderr, "Failed to open rtlsdr device #%d.\n", dev_index);
		exit(1);
	}

	fprintf(stderr, "\n");

	r = rtlsdr_read_eeprom(dev, buf, 0, EEPROM_SIZE);
	if (r < 0) {
		if (r == -3)
			fprintf(stderr, "No EEPROM has been found.\n");
		else
			fprintf(stderr, "Failed to read EEPROM, err %i.\n", r);
		goto exit;
	}

	if (r < 0)
		return -1;

	fprintf(stderr, "Current configuration:\n");
	parse_eeprom_to_conf(&conf, buf);
	dump_config(&conf);

	if (filename) {
		file = fopen(filename, flash_file ? "rb" : "wb");
		if (!file) {
			fprintf(stderr, "Error opening file!\n");
			goto exit;
		}
		if (flash_file) {
			if (fread(buf, 1, sizeof(buf), file) != sizeof(buf))
				fprintf(stderr, "Error reading file!\n");
		} else {
			if (fwrite(buf, 1, sizeof(buf), file) != sizeof(buf))
				fprintf(stderr, "Short write, exiting!\n");
			else
				fprintf(stderr, "\nDump to %s successful.\n", filename);
		}
	}

	if (manuf_str)
		strncpy((char*)&conf.manufacturer, manuf_str, MAX_STR_SIZE);

	if (product_str)
		strncpy((char*)&conf.product, product_str, MAX_STR_SIZE);

	if (serial_str) {
		conf.have_serial = 1;
		strncpy((char*)&conf.serial, serial_str, MAX_STR_SIZE);
	}

	if (ir_endpoint != 0)
		 conf.enable_ir = (ir_endpoint > 0) ? 1 : 0;

	if (!change)
		goto exit;

	fprintf(stderr, "\nNew configuration:\n");

	if (default_config != CONF_NONE)
		gen_default_conf(&conf, default_config);

	if (!flash_file) {
		if (gen_eeprom_from_conf(&conf, buf) < 0)
			goto exit;
	}

	parse_eeprom_to_conf(&conf, buf);
	dump_config(&conf);

	fprintf(stderr, "Write new configuration to device [y/n]? ");

	while ((ch = getchar())) {
		if (ch != 'y')
			goto exit;
		else
			break;
	}

	r = rtlsdr_write_eeprom(dev, buf, 0, flash_file ? EEPROM_SIZE : 128);
	if (r < 0)
		fprintf(stderr, "Error while writing EEPROM: %i\n", r);
	else
		fprintf(stderr, "\nConfiguration successfully written.\n"
				"Please replug the device for changes"
				" to take effect.\n");

exit:
	if (file)
		fclose(file);

	rtlsdr_close(dev);

	return r >= 0 ? r : -r;
}
